Introduce metrics

achaikou opened the pull request:
Preparations for supporting `traceheader_count` > 1.

This parameter would have been added in every place where `trace0` and `trace_bsize` are used.
After this PR all 3 parameters can be removed from the functions and will be treated the same way as `elemsize` currently is - set values would be read from `segy_datasource`.

To make code more consistent, endianness and encoding would now be treated the same way as all other parameters. With relation to this deducing endianness from file was implemented too.

achaikou committed:
Do not recover on bad format

It is unknown why special care was taken to deal with incorrectly set
format in binary header. Decision seems to stem from #241 but no
justification is given.
Currently we have no knowledge of files with bad format set.

We do not allow user to override other fields, like sample count and
extended text headers, so it is unclear why format is so special.

Lets assume we update the test that is supposed to check we warn user
when format is bad and that we use what we assume is the default one.
Update it by reading the trace.

def test_missing_format_ibmfloat_fallback(small):
    with segyio.open(small, mode = 'r+') as f:
        val = f.trace[0][0]
        f.bin[segyio.su.format] = 0

    with segyio.open(small) as f:
        assert int(f.format) == 1
        assert f.dtype       == np.dtype(np.float32)

        assert f.trace[0][0] == val

It fails on the old code.

As we can't even read trace correctly when we guess the format and we
didn't get lots of complains, it likely means that
files with not-set format is not a problem.

So we should be fine with dropping the support altogether.

achaikou committed:
Turn segy_set_endianness to segy_endianness

Function would be responsible only for retrieving endianness, not for
setting the value on segy_datasource.

Setting the value becomes responsibility of the caller of this function.

achaikou committed:
Change default endianness from big to file defined

Should not severely affect users as "0", which is likely the value in
the binary field, means "big endian".

achaikou committed:
Turn segy_set_encoding to segy_encoding

Function would be responsible only for retrieving encoding, not for
setting the value on segy_datasource.

Setting the value becomes responsibility of the caller of this function.

achaikou committed:
Remove segy_set_format

Setting the elemsize becomes responsibility of the caller.

achaikou committed:
Update field limit test files for linetrc/reeltrc

Before we assumed that TRACE_SEQUENCE_LINE and TRACE_SEQUENCE_FILE to be
signed ints.
But recently we changed them to be of unsigned type (linetrc/reeltrc)
and did not update relevant test files.

achaikou committed:
Swap endianness for all binary header fields

Automatically collect binary header fields based on the fields map.

To test the changes, increment-decrement files got flipped via
flip-endianness. Endianness int constant was set to allow future
automatic endianness recognition.

f3.sgy is changed because revision was flipped incorrectly before (those
are two 1-byte fields, not one 2-byte).

achaikou committed:
Declare C++11 as used in whole project

ajaust reviewed (commented):
This looks great overall. 🎉 

I just left some minor comments for clarification. The only (potentially) major thing is the naming, i.e., "Is there a better name than `metrics`" for the properties of the file. I have to think a bit more about this since I have no answer myself.

achaikou commented on python/segyio/segyio.cpp:
```diff
@@ -641,46 +641,28 @@ PyObject* segyopen( segyfd* self, PyObject* args, PyObject* kwargs ) {
-                                 "no data traces past headers" );
+       case SEGY_TRACE_SIZE_MISMATCH:
+           return RuntimeError( "trace count inconsistent with file size, "
+                                "trace lengths possibly of non-uniform" );
```
Will not fix because I remove this error message altogether in a later commit anyway...

Those error messages help very little anyway.

achaikou commented on applications/segyio-catr.c:
```diff
@@ -699,29 +699,35 @@ int main( int argc, char** argv ) {
                         break;
+                    case SEGY_TR_SEQ_LINE:
+                    case SEGY_TR_SEQ_FILE:
+                        f = (long)((uint32_t)buf);
```
This was done on purpose by me trying to make one failing test pass :smile: 
I was satisfied with the first line that worked :smile: 

I believe here I can use `buf` as it contains just 4 bytes. I state that I want to interpret those bytes as uint32.

As I translate 4 bytes to 4 bytes, I am sure no bytes are dropped.
I like that more than uint16f situation, because there, not matter whether I choose 8-bytes `f` or 4-bytes `buf`, some bytes will be dropped to make the value 2-bytes.

But I think I should cast to `long long` instead of long. I switched to long long as it is certain to have 8 bytes and we need a type that can hold both uint32 and int32 values.

yngve793 commented on python/segyio/segyio.cpp:
```diff
@@ -755,42 +737,8 @@ PyObject* segycreate( segyfd* self, PyObject* args, PyObject* kwargs ) {
-            assert(false || "format should already be checked" );
-            break;
-    }
+    segy_set_format( ds, format );
```
Even better :-)

achaikou reviewed (commented):
Thanks for the review!

yngve793 commented on lib/src/segy.c:
```diff
@@ -1449,16 +1449,29 @@ int segy_set_format( segy_datasource* ds, int format ) {
+    if( err != SEGY_OK ) return err;
+
+    segy_field_data fd;
+    segy_get_binfield( binheader, SEGY_BIN_INTEGER_CONSTANT, &fd );
```
Happy to see SEGY_BIN_INTEGER_CONSTANT getting used.

yngve793 reviewed (approved):
Very nice solution. :-)

achaikou commented on lib/include/segyio/segy.h:
```diff
@@ -443,16 +443,12 @@ int segy_format( const char* binheader );
 int segy_set_format( segy_datasource*, int format );
 
-/* set file as LSB/MSB (little/big endian)
+/* Find out whether file is LSB/MSB (little/big endian).
```
Not done because I remove `segy_set_format` completely in one of the next commits.

achaikou commented on python/segyio/create.py:
```diff
@@ -6,6 +6,8 @@
 from .utils import (
     FileDatasourceDescriptor,
     StreamDatasourceDescriptor,
+    c_endianness,
```
Was not introduced in this commit but earlier.

There are two representations of "endianness" and "encoding" properties - one in Python, another in C.
`c_endianness` and `c_encoding` functions are responsible to translate value for encoding/endianness in python ("ebcdic", "little") to corresponding value in C (0 or 1).
I think I named them so because `python_endianness_to_c_endianness` sounded too long.

Could change the name if you suggest what to change it to. :slightly_smiling_face:

yngve793 replied:
Fixed.

achaikou commented on applications/segyio-catr.c:
```diff
@@ -470,7 +470,6 @@ static struct options parse_options( int argc, char** argv ){
     static struct option long_options[] = {
         { "trace",          required_argument,  0, 't' },
         { "range",          required_argument,  0, 'r' },
-        { "format",         required_argument,  0, 'f' },
```
This whole change is based on PO being unaware of existing files with wrong format value set.
In the issue where this code was introduced there was also no talk about this.

So I too hope this is really not a valid case anymore and as you say, users could use previous versions, if they have to  :+1:

achaikou committed:
Drop ds->lsb

achaikou reviewed (commented):
:heart:

yngve793 reviewed (approved):
Looks good.

ajaust reviewed (approved):
The changes look good to me. The fixups are asle nice. 👍 

All my comments are resolved. I just left one open for reference since Yngve commented on the same thing (renaming `metrics` to `metadata`).

achaikou committed:
Make formatsize public

achaikou committed:
Move endianness and encoding out of init

Those are not features of opened datasource, but features of SEG-Y file,
so they fit better to segyopen/segycreate/suopen.
It also allows us to control whether we want to retrieve them or assign
them.

achaikou committed:
Collect metadata

Collect all basic metadata required for future calculations.

Some code is updated to use the new function, some is not. Both old and
new approach is valid, so there is no need to update everything.

achaikou committed:
Drop ds->encoding

achaikou committed:
Drop ds->elemsize

achaikou committed:
Rename c_encoding and c_endianness
